src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -407,23 +407,64 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
             }
         }
 
-        [RequiresUnreferencedCode("In case type is a Nullable<T>, cannot statically analyze what the underlying type is so its members may be trimmed.")]
-        private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type type)
+        [RequiresUnreferencedCode(
+            "In case type is a Nullable<T>, cannot statically analyze what the underlying type is so its members may be trimmed.")]
+        private static object CreateInstance(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors |
+                                        DynamicallyAccessedMemberTypes.NonPublicConstructors)]
+            Type type,
+            IConfiguration config,
+            BinderOptions options)
         {
             Debug.Assert(!type.IsArray);
 
-            if (type.IsAbstract)
+            if (type.IsInterface || type.IsAbstract)
             {
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+            bool hasParameterlessConstructor =
+                type.IsValueType || constructors.Any(ctor => ctor.GetParameters().Length == 0);
+
+            if (!type.IsValueType && constructors.Length == 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.Error_MissingPublicInstanceConstructor, type));
+            }
+
+            if (constructors.Length > 1 && !hasParameterlessConstructor)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                throw new InvalidOperationException(SR.Format(SR.Error_MultipleParameterizedConstructors, type));
+            }
+
+            if (constructors.Length == 1 && !hasParameterlessConstructor)
+            {
+                ConstructorInfo constructor = constructors[0];
+                ParameterInfo[] parameters = constructor.GetParameters();
+
+                if (!CanBindToTheseConstructorParameters(parameters, out string nameOfInvalidParameter))
+                {
+                    throw new InvalidOperationException(SR.Format(SR.Error_CannotBindToConstructorParameter, type, nameOfInvalidParameter));
+                }
+
+
+                List<PropertyInfo> properties = GetAllProperties(type);
+                List<FieldInfo> fields = GetAllFields(type);
```

Comment:
`fields` is no longer used. And with that, `GetAllFields` can be removed as well.